v2/api_error.go

```diff
@@ -0,0 +1,90 @@
+package gax
+
+// code snippet credit to @ndietz
+import (
+	"encoding/json"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/status"
+)
+
+//hold the google/rpc/error_details.proto messages.
+type ErrDetails struct {
+	BadRequest          *errdetails.BadRequest          `json:",omitempty"`
+	PreconditionFailure *errdetails.PreconditionFailure `json:",omitempty"`
+	QuotaFailure        *errdetails.QuotaFailure        `json:",omitempty"`
+	RetryInfo           *errdetails.RetryInfo           `json:",omitempty"`
+	ResourceInfo        *errdetails.ResourceInfo        `json:",omitempty"`
+	RequestInfo         *errdetails.RequestInfo         `json:",omitempty"`
+	DebugInfo           *errdetails.DebugInfo           `json:",omitempty"`
+	Help                *errdetails.Help                `json:",omitempty"`
+	LocalizedMesage     *errdetails.LocalizedMessage    `json:",omitempty"`
+
+	//store unidentifiable error details
+	Unknown []interface{} `json:",omitempty"`
+}
+
+type APIError struct {
+	err     error
+	status  *status.Status
+	details ErrDetails
+}
+
+func (a *APIError) Details() ErrDetails {
+	return a.details
+}
+
+func (a *APIError) Unwrap() error {
+	return a.err
+}
+
+func (a *APIError) Error() string {
+
+	strr, _ := json.Marshal(a.details)
+	return a.err.Error() + "\n" + "Here are the details: " + "\n" + string(strr)
+
+}
+
+func (a *APIError) GRPCStatus() *status.Status {
+	return a.status
+}
+
+func FromError(err error) (*APIError, bool) {
+	if err == nil {
+		return nil, false
+	}
+	msg := ErrDetails{}
```

Comment:
`msg` is kind of confuse here, because we are signing the error details value here. Might need to change errDetails.